torchtune/training/_distributed.py

```diff
@@ -726,3 +751,112 @@ def prepare_mha_for_tp(
+):
+    """
+    Creates a context manager to confine to flash/efficient/cuDNN attention backends.
+
+    Returns:
+        A context manager function that takes an optional context parallel context.
+    """
+
+    @contextlib.contextmanager
+    def context(cp_context: Union[Generator[None, None, None], None] = None):
+        with contextlib.ExitStack() as stack:
+            if cp_context is not None:
+                stack.enter_context(
+                    sdpa_kernel(
+                        [
+                            SDPBackend.FLASH_ATTENTION,
+                            SDPBackend.EFFICIENT_ATTENTION,
+                            SDPBackend.CUDNN_ATTENTION,
+                        ]
+                    )
+                )
+                stack.enter_context(cp_context)
+
+            yield
+
+    return context
+
+
+def get_context_parallel_context(
+    *,
+    cp_enabled: bool = False,
+    world_mesh: torch.distributed.DeviceMesh,
+    model: TransformerDecoder,
+    model_inputs: list[torch.Tensor],
+) -> Generator[None, None, None]:
+    """
+    Context manager for applying context parallelism to a model. In addition to applying the
+    standard context manager to patch SDPA and shard model inputs and buffers along the sequence
+    dimension, this context manager also calls into _get_sdpa_context to filter to acceptable SDPA backends.
+
+    Args:
+        cp_enabled (bool): Whether context parallel is enabled. Default: False
+        world_mesh (torch.distributed.DeviceMesh): Global device mesh.
+        model (TransformerDecoder): Model to apply context parallelism to.
+        model_inputs (list[torch.Tensor]): List of any model inputs which should be
+            sharded along sequence dimension.
+
+    Returns:
+        A context manager applying context parallelism if cp_enabled is True. Otherwise a context manager
+        disabling the math SDPA backend.
+
+    Raises:
+        ValueError: if cp_enabled is True but world_mesh does not contain a "cp" dimension
+
+    Example:
+        ```python
+        batch = {"inputs": inputs, "labels": labels}
+        with get_context_parallel_context(
+            cp_enabled=parallel_dims.cp_enabled,
+            cp_mesh=world_mesh["cp"] if parallel_dims.cp_enabled else None,
+            model_inputs=list(batch.values()),
+            model_buffers=model.buffers(),
+        ):
+            logits = model(inputs)
+            loss = loss(logits, labels)
+            loss.backward()
+        ```
+    """
+
+    if cp_enabled and "cp" not in world_mesh.mesh_dim_names:
+        raise ValueError(
+            "Context parallel is enabled but no context parallel device mesh is provided."
+        )
+    # TODO: context parallel for multimodal models requires extra work
+    if cp_enabled and not isinstance(model, TransformerDecoder):
+        raise ValueError(
+            "Context parallel not supported for models other than TransformerDecoder"
```

Comment:
I think this error message should just state that "Only text models are supported" as a user might not understand why they're getting this error.